worker/provisioner/provisioner_task.go

```diff
@@ -689,101 +715,358 @@ func (task *provisionerTask) maintainMachines(machines []*apiprovisioner.Machine
+	switch {
+	case b[i].machineIds.Size() < b[j].machineIds.Size():
+		return true
+	case b[i].machineIds.Size() == b[j].machineIds.Size():
+		return b[i].zoneName < b[j].zoneName
+	}
+	return false
+}
+
+func (b byPopulationThenName) Swap(i, j int) {
+	b[i], b[j] = b[j], b[i]
+}
+
+// startMachines starts a goroutine for each specified machine to
+// start it.  The specified machines are provided the name of the
+// first availability zone to try.  Errors from individual start
+// machine attempts will be logged.
+func (task *provisionerTask) startMachines(machines []*apiprovisioner.Machine) error {
+	if len(machines) == 0 {
+		return nil
+	}
+
+	// Get the distributionGroups for each machine now to avoid
+	// successive calls to DistributionGroupByMachineId which will
+	// return the same data.
+	machineTags := make([]names.MachineTag, len(machines))
+	for i, machine := range machines {
+		machineTags[i] = machine.MachineTag()
+	}
+	machineDistributionGroups, err := task.distributionGroupFinder.DistributionGroupByMachineId(machineTags...)
+	if err != nil {
+		return err
+	}
+
+	wg := sync.WaitGroup{}
+	wg.Add(len(machines))
+	errMachines := make([]error, len(machines))
+
+	for i, m := range machines {
+		// Make sure we shouldn't be stopping before we start the next machine
+		select {
+		case <-task.catacomb.Dying():
+			return task.catacomb.ErrDying()
+		default:
 		}
 
-		if err := task.startMachine(m, pInfo, startInstanceParams); err != nil {
-			return errors.Annotatef(err, "cannot start machine %v", m)
+		if machineDistributionGroups[i].Err != nil {
+			return err
+		}
+
+		go func(machine *apiprovisioner.Machine, dg []string, index int) {
+			defer wg.Done()
+			pInfo, startInstanceParams, err := task.setupToStartMachine(machine)
+			if err != nil {
+				errMachines[index] = err
+				task.removeMachinesFromAZMap(machine)
+				return
+			}
+			if err := task.startMachine(machine, pInfo, startInstanceParams, dg); err != nil {
+				errMachines[index] = err
+				task.removeMachinesFromAZMap(machine)
+			}
+		}(m, machineDistributionGroups[i].MachineIds, i)
+	}
+
+	// ToDo ProvisionerParallelization 2017-10-09
+	// Is this necessary?
+	wg.Wait()
+	for i, err := range errMachines {
+		if err != nil {
+			task.setErrorStatus("cannot start machine %s: %v", machines[i], err)
 		}
 	}
 	return nil
 }
 
 func (task *provisionerTask) setErrorStatus(message string, machine *apiprovisioner.Machine, err error) error {
 	logger.Errorf(message, machine, err)
-	if err := machine.SetInstanceStatus(status.ProvisioningError, err.Error(), nil); err != nil {
+	if err2 := machine.SetInstanceStatus(status.ProvisioningError, err.Error(), nil); err2 != nil {
 		// Something is wrong with this machine, better report it back.
-		return errors.Annotatef(err, "cannot set error status for machine %q", machine)
+		return errors.Annotatef(err2, "cannot set error status for machine %q", machine)
 	}
-	return nil
+	return err
```

Comment:
sorry I missed this before. please don't do this, it's too surprising. going by the name of the method, its purpose is to set the error status. it should only return an error if it fails to set the error status.

Reply:
set back to nil, added setInstanceStatusAndRemoveMachineAZ() to help other problems.